images/cfw/opt/x1plus/lib/python/x1plus/services/x1plusd/modules/i2c_drivers/tca8418_driver.py

```diff
@@ -0,0 +1,214 @@
+            buttons[f"btn_{x}"] = "pressed" if x in self.currently_pressed else "released"
+        return buttons
+
+
+    async def _task(self):
+       
+        # Init MQTT structure
+        await self.sensors.publish(self.name, type = name, rows = self.rows, columns = self.cols, **self.get_states() )
+
+        while True:
+            updated = False
+            try:
+
+                int_stat = self.read_register(self.REG_INT_STAT)[0]
+                
+                if int_stat not in [self.INT_STAT_GPI_INT, self.INT_STAT_K_INT]:
+                    # No events in FIFO are key events
+                    self.write_register(self.REG_INT_STAT, 0xFF)
+                    await asyncio.sleep(self.interval_ms / 1000.0)
+                    continue
+
+                events_count = self.read_register(self.REG_KEY_LCK_EC)[0] & 0x0F
+
+                for _ in range(events_count):
+                    event_data = self.read_register(self.REG_KEY_EVENT_A, 1)[0]
+
+                    is_release = event_data & 0x80 == 0
+
+                    # Key always comes back as (row * 10) + column, row being 0 indexed, column 1 indexed in formula
+                    # I.E., Key 39 = Row 4, Column 9. Key 40 = Row 4, Column 10. Key 41 = Row 5, Column 1
+                    key = int(event_data & 0x7F)
+                    row = (key - 1) // 10
+                    col = (key - 1) % 10
+                    key = (row * self.cols) + col + 1
+
+                    if key < 1 or key > 80 or key > (self.rows * self.cols):
+                        continue
+
+                    if is_release:
+                        logger.debug(f"{name.upper()} key: {key} was released")
+                        self.currently_pressed.discard(key)
+                    else:
+                        if key not in self.currently_pressed:
+                            logger.debug(f"{name.upper()} key: {key} was pressed")
+                            self.currently_pressed.add(key)
+                        if action := self.mapped_actions.get(str(key), None):
+                            try: 
+                                await self.actions.execute(action)
+                            except Exception as e:
+                                logger.error(f"Error performing {name.upper()} key-press action for key: {key}, action {action}: {e.__class__.__name__}: {e}")
+                                await self.sensors.publish(self.name, type = name, inop = { 'exception': f"{e.__class__.__name__}: {e}" })   
+                    updated = True
+                  
+                self.write_register(self.REG_INT_STAT, 0xFF)          
+ 
+            except Exception as e:
+                logger.error(f"Error reading {name.upper()} buttons: {e.__class__.__name__}: {e}")
+                await self.sensors.publish(self.name, type = name, inop = { 'exception': f"{e.__class__.__name__}: {e}" })
+                # In case of recurring error, do not spam log/system. Sleep 5s
```

Comment:
TODO: Fail out of driver entirely if recurrent error - one offs can/should be logged. Recurring errors in loop are rare and would only occur in failure to read, so perhaps if this is hit 5 times in X timeframe (remove sleep) we disconnect driver